geo/src/algorithm/bool_ops/i_overlay_integration.rs

```diff
@@ -0,0 +1,299 @@
+use crate::geometry::Coord;
+use crate::GeoNum;
+use i_overlay::core::fill_rule::FillRule;
+use i_overlay::core::overlay::ShapeType;
+use i_overlay::core::overlay_rule::OverlayRule;
+use i_overlay::string::clip::ClipRule;
+
+pub trait BoolOpsCoord<T>: Copy {
+    fn new(x: T, y: T) -> Self;
+    fn x(&self) -> T;
+    fn y(&self) -> T;
+}
+
+/// A geometry coordinate number suitable for performing geometric boolean operations.
+pub trait BoolOpsNum: GeoNum {
+    type CoordType: BoolOpsCoord<Self>;
+    type OverlayType: BoolOpsOverlay<CoordType = Self::CoordType>;
+    type StringOverlayType: BoolOpsStringOverlay<CoordType = Self::CoordType>;
+
+    fn to_bops_coord(geo_coord: Coord<Self>) -> Self::CoordType {
+        Self::CoordType::new(geo_coord.x, geo_coord.y)
+    }
+
+    fn to_geo_coord(bops_coord: Self::CoordType) -> Coord<Self> {
+        Coord {
+            x: bops_coord.x(),
+            y: bops_coord.y(),
+        }
+    }
+}
+
+pub trait BoolOpsOverlay {
+    type CoordType;
+    type OverlayGraph: BoolOpsOverlayGraph<CoordType = Self::CoordType>;
+    fn new() -> Self;
+    fn add_path(&mut self, path: Vec<Self::CoordType>, shape_type: ShapeType);
+    fn into_graph(self, fill_rule: FillRule) -> Self::OverlayGraph;
+}
+
+pub(super) trait BoolOpsOverlayGraph {
+    type CoordType;
+    fn extract_shapes(&self, overlay_rule: OverlayRule) -> Vec<Vec<Vec<Self::CoordType>>>;
+}
+
+pub trait BoolOpsStringOverlay {
+    type CoordType;
+    type StringGraph: BoolOpsStringGraph<CoordType = Self::CoordType>;
+    fn new() -> Self;
+    fn add_shape_path(&mut self, path: Vec<Self::CoordType>);
+    fn add_string_line(&mut self, path: [Self::CoordType; 2]);
+    fn into_graph(self, fill_rule: FillRule) -> Self::StringGraph;
+}
+
+pub(super) trait BoolOpsStringGraph {
+    type CoordType;
+    fn clip_string_lines(&self, clip_rule: ClipRule) -> Vec<Vec<Self::CoordType>>;
+}
+
+mod f64 {
+    use super::{ClipRule, FillRule, OverlayRule, ShapeType};
+    use i_overlay::f64::{
+        graph::F64OverlayGraph,
+        overlay::F64Overlay,
+        string::{F64StringGraph, F64StringOverlay},
+    };
+    use i_overlay::i_float::f64_point::F64Point;
+
+    impl super::BoolOpsNum for f64 {
+        type CoordType = F64Point;
+        type OverlayType = F64Overlay;
+        type StringOverlayType = F64StringOverlay;
+    }
+
+    impl super::BoolOpsCoord<f64> for F64Point {
```

Comment:
Maybe inlining all of these would be a good idea to get a bit more performance

Reply:
Typically I don't like to back seat drive the compiler, but there are some nice boosts here. 

re: https://std-dev-guide.rust-lang.org/policy/inline.html

Maybe this falls under the "public, small, non-generic functions."

<details>
<summary>cargo bench from inline</summary>

```
Circular polygon boolean-ops/bops::intersection/256                                                                                            
                        time:   [456.98 µs 461.66 µs 467.85 µs]                                                                                
                        change: [-1.5757% -0.8205% +0.0412%] (p = 0.11 > 0.05)                                                                 
                        No change in performance detected.                                                                                     
Found 1 outliers among 10 measurements (10.00%)                                                                                                
  1 (10.00%) high severe                                               
Circular polygon boolean-ops/bops::union/256                           
                        time:   [457.72 µs 458.64 µs 459.76 µs]                                                                                
                        change: [-2.0602% -1.7872% -1.5125%] (p = 0.00 < 0.05)                                                                 
                        Performance has improved.                                                                                              
Circular polygon boolean-ops/bops::intersection/512                    
                        time:   [1.3860 ms 1.3888 ms 1.3916 ms]                                                                                
                        change: [-10.975% -10.623% -10.346%] (p = 0.00 < 0.05)                                                                 
                        Performance has improved.                                                                                              
Circular polygon boolean-ops/bops::union/512                                                                                                   
                        time:   [1.3902 ms 1.3922 ms 1.3943 ms]                                                                                
                        change: [-11.574% -10.967% -10.469%] (p = 0.00 < 0.05)                                                                 
                        Performance has improved.                                                                                              
Found 1 outliers among 10 measurements (10.00%)                                                                                                
  1 (10.00%) high mild                                                                                                                         
Circular polygon boolean-ops/bops::intersection/1024                                                                                           
                        time:   [4.2887 ms 4.2972 ms 4.3112 ms]                                                                                
                        change: [-8.7721% -7.8505% -7.1175%] (p = 0.00 < 0.05)                                                                 
                        Performance has improved.                                                                                              
Found 2 outliers among 10 measurements (20.00%)                        
  1 (10.00%) high mild                                                                                                                         
  1 (10.00%) high severe                                               
Circular polygon boolean-ops/bops::union/1024                                                                                                  
                        time:   [4.2956 ms 4.3075 ms 4.3290 ms]                                                                                
                        change: [-7.9012% -7.3024% -6.5849%] (p = 0.00 < 0.05)                                                                 
                        Performance has improved.                                                                                              
Found 2 outliers among 10 measurements (20.00%)                        
  1 (10.00%) high mild                                                                                                                         
  1 (10.00%) high severe                                                                                                                       
Circular polygon boolean-ops/bops::intersection/2048                                                                                           
                        time:   [14.081 ms 14.098 ms 14.122 ms]                                                                                                                                                                                                                                
                        change: [+0.1742% +0.4568% +0.8040%] (p = 0.01 < 0.05)                                                                 
                        Change within noise threshold.                                                                                         
Found 1 outliers among 10 measurements (10.00%)                                                                                                
  1 (10.00%) high severe                                                                                                                                                                                                                                                                       
Circular polygon boolean-ops/bops::union/2048                          
                        time:   [14.078 ms 14.116 ms 14.160 ms]                                                                                                                                                                                                                                
                        change: [-0.2573% +0.3188% +0.8433%] (p = 0.30 > 0.05)                                                                 
                        No change in performance detected.                                                                                     
Circular polygon boolean-ops/bops::intersection/4096                                                                                           
                        time:   [50.853 ms 50.954 ms 51.038 ms]                                                                                                                                                                                                                                
                        change: [-0.8439% -0.2164% +0.4735%] (p = 0.54 > 0.05)                                                                 
                        No change in performance detected.                                                                                     
Found 2 outliers among 10 measurements (20.00%)                                                                                                                                                                                                                                                
  1 (10.00%) high mild                                                 
  1 (10.00%) high severe                                                                                                                                                                                                                                                                       
Circular polygon boolean-ops/bops::union/4096                          
                        time:   [50.889 ms 50.969 ms 51.048 ms]                                                                                                                                                                                                                                
                        change: [-0.7440% -0.2436% +0.3009%] (p = 0.40 > 0.05)                                                                 
                        No change in performance detected.                                                                                                                                                                                                                                     
Found 2 outliers among 10 measurements (20.00%)                        
  2 (20.00%) high severe                                                                                                                       
Circular polygon boolean-ops/bops::intersection/8192                                                                                           
                        time:   [188.02 ms 188.28 ms 188.58 ms]                                                                                                                                                                                                                                
                        change: [-1.2461% -1.0530% -0.8445%] (p = 0.00 < 0.05)                                                                 
                        Change within noise threshold.                                                                                                                                                                                                                                         
Circular polygon boolean-ops/bops::union/8192                          
                        time:   [187.91 ms 188.13 ms 188.39 ms]                                                                                                                                                                                                                                
                        change: [-1.1870% -1.0354% -0.8655%] (p = 0.00 < 0.05)                                                                 
                        Change within noise threshold.    
```

</details>